gdnative-core/src/class.rs

```diff
@@ -60,71 +69,148 @@ pub trait NativeClassMethods: NativeClass {
+            (gd_api.godot_method_bind_ptrcall)(set_library, native_script, args.as_mut_ptr(), std::ptr::null_mut());
+
+            let this = T::Base::construct();
 
-            Self {
-                this: self.this,
-                _marker: PhantomData,
+            assert_ne!(std::ptr::null_mut(), this.to_sys(), "base object should not be null");
+
+            let mut args: [*const libc::c_void; 1] = [native_script as *const _];
+            (gd_api.godot_method_bind_ptrcall)(object_set_script, this.to_sys(), args.as_mut_ptr(), std::ptr::null_mut());
+
+            let script_ptr = (gd_api.godot_nativescript_get_userdata)(this.to_sys()) as *const libc::c_void;
+
+            assert_ne!(std::ptr::null(), script_ptr, "script instance should not be null");
+
+            let script = T::UserData::clone_from_user_data_unchecked(script_ptr);
+
+            object::unref(native_script);
+
+            Instance {
+                this,
+                script,
             }
         }
     }
 
-    fn get_impl(&self) -> &RefCell<T> {
-        unsafe {
-            let api = get_api();
-            let ud = (api.godot_nativescript_get_userdata)(self.this);
-            &*(ud as *const _ as *const RefCell<T>)
-        }
+    pub fn into_base(self) -> T::Base {
+        self.this
     }
 
-    #[doc(hidden)]
-    pub unsafe fn sys(&self) -> *mut sys::godot_object {
-        self.this
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used on reference counted types for multiple times.
+    pub fn map<F, U>(&self, op: F) -> Result<U, <T::UserData as Map>::Err>
+    where
+        T::Base: Clone,
+        T::UserData: Map,
+        F: FnOnce(&T, T::Base) -> U,
+    {
+        self.script.map(|script| op(script, self.this.clone()))
+    }
+
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used on reference counted types for multiple times.
+    pub fn map_mut<F, U>(&self, op: F) -> Result<U, <T::UserData as MapMut>::Err>
+    where
+        T::Base: Clone,
+        T::UserData: MapMut,
+        F: FnOnce(&mut T, T::Base) -> U,
+    {
+        self.script.map_mut(|script| op(script, self.this.clone()))
+    }
+
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used for multiple times via aliasing. For reference-counted types
+    /// behaves like the safe `map`, which should be preferred.
+    pub unsafe fn map_aliased<F, U>(&self, op: F) -> Result<U, <T::UserData as Map>::Err>
+    where
+        T::UserData: Map,
+        F: FnOnce(&T, T::Base) -> U,
+    {
+        self.script.map(|script| op(script, T::Base::from_sys(self.this.to_sys())))
+    }
+
+    /// Calls a function with a NativeClass instance and its owner, and returns its return
+    /// value. Can be used for multiple times via aliasing. For reference-counted types
```

Comment:
Is this notice still true when considering that this allows multiple mutable aliases?

Reply:
You can't make multiple mutable aliases of the `NativeClass` because the reference is only valid for the closure. Even if you manage to call `map_mut` again inside, the user data wrapper should prevent it.

The `_aliased` part refers to the Godot object, and for Reference types `from_sys(to_sys())` is indeed (currently) equivalent to a clone.